File: haskell/glue/src/Glue/Lib/Math/Power/Sqrt.hs
module Glue.Lib.Math.Power.Sqrt where

import Glue.Eval (Eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..))

-- Square root function
-- Mirrors Haskell Glue.Lib.Math.Power.Sqrt.sqrt exactly
sqrt :: IR Eval
sqrt = NativeFunc sqrtImpl

-- Square root function implementation
-- Mirrors Haskell Glue.Lib.Math.Power.Sqrt.sqrtImpl exactly
sqrtImpl :: IR Eval -> Eval (IR Eval)
sqrtImpl arg = case arg of
    Integer n -> pure $ Float (Prelude.sqrt (fromIntegral n))
    Float n -> pure $ Float (Prelude.sqrt n)
    _ -> throwError $ wrongArgumentType ["number"]

File: dart/glue/lib/src/lib/math/power/sqrt.dart
import 'dart:math' as math;

import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Square root function
/// Mirrors Haskell Glue.Lib.Math.Power.Sqrt.sqrt exactly
final Ir sqrt = IrNativeFunc(sqrtImpl);

/// Square root function implementation
/// Mirrors Haskell Glue.Lib.Math.Power.Sqrt.sqrtImpl exactly
Eval<Ir> sqrtImpl(Ir arg) {
  return switch (arg) {
    IrInteger(value: final n) => Eval.pure(IrFloat(math.sqrt(n.toDouble()))),
    IrFloat(value: final n) => Eval.pure(IrFloat(math.sqrt(n))),
    _ => throwError(wrongArgumentType(['number'])),
  };
}
